File: src/Compilers/Test/Utilities/VisualBasic/SyntaxTreeExtensions.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Microsoft.CodeAnalysis.PooledObjects
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Public Module SyntaxTreeExtensions
    <System.Runtime.CompilerServices.Extension()>
    Public Function WithReplace(syntaxTree As SyntaxTree, offset As Integer, length As Integer, newText As String) As SyntaxTree
        Dim oldFullText = syntaxTree.GetText()
        Dim newFullText = oldFullText.WithChanges(New TextChange(New TextSpan(offset, length), newText))
        Return syntaxTree.WithChangedText(newFullText)
    End Function

    <System.Runtime.CompilerServices.Extension()>
    Friend Function WithReplaceFirst(syntaxTree As SyntaxTree, oldText As String, newText As String) As SyntaxTree
        Dim oldFullText = syntaxTree.GetText().ToString()
        Dim offset As Integer = oldFullText.IndexOf(oldText, StringComparison.Ordinal)
        Dim length As Integer = oldText.Length
        Return WithReplace(syntaxTree, offset, length, newText)
    End Function

    <System.Runtime.CompilerServices.Extension()>
    Public Function WithReplace(syntaxTree As SyntaxTree, startIndex As Integer, oldText As String, newText As String) As SyntaxTree
        Dim oldFullText = syntaxTree.GetText().ToString()
        Dim offset As Integer = oldFullText.IndexOf(oldText, startIndex, StringComparison.Ordinal)
        Dim length As Integer = oldText.Length
        Return WithReplace(syntaxTree, offset, length, newText)
    End Function

    <System.Runtime.CompilerServices.Extension()>
    Public Function WithInsertAt(syntaxTree As SyntaxTree, offset As Integer, newText As String) As SyntaxTree
        Return WithReplace(syntaxTree, offset, 0, newText)
    End Function

    <System.Runtime.CompilerServices.Extension()>
    Public Function WithInsertBefore(syntaxTree As SyntaxTree, existingText As String, newText As String) As SyntaxTree
        Dim oldFullText = syntaxTree.GetText().ToString()
        Dim offset As Integer = oldFullText.IndexOf(existingText, StringComparison.Ordinal)
        Return WithReplace(syntaxTree, offset, 0, newText)
    End Function

    <System.Runtime.CompilerServices.Extension()>
    Public Function WithRemoveAt(syntaxTree As SyntaxTree, offset As Integer, length As Integer) As SyntaxTree
        Return WithReplace(syntaxTree, offset, length, String.Empty)
    End Function

    <System.Runtime.CompilerServices.Extension()>
    Public Function WithRemoveFirst(syntaxTree As SyntaxTree, oldText As String) As SyntaxTree
        Return WithReplaceFirst(syntaxTree, oldText, String.Empty)
    End Function

    <System.Runtime.CompilerServices.Extension()>
    Friend Function Dump(node As SyntaxNode) As String
        Dim visitor = New VisualBasicSyntaxPrinter()
        visitor.Visit(node)
        Return visitor.Dump()
    End Function

    <System.Runtime.CompilerServices.Extension()>
    Friend Function Dump(node As SyntaxTree) As String
        Return node.GetRoot().Dump()
    End Function

    Private Class VisualBasicSyntaxPrinter
        Inherits VisualBasicSyntaxWalker

        ReadOnly Dim builder As PooledStringBuilder
        Dim indent As Integer = 0

        Sub New()
            builder = PooledStringBuilder.GetInstance()
        End Sub

        Function Dump() As String
            Return builder.ToStringAndFree()
        End Function

        Public Overrides Sub DefaultVisit(node As SyntaxNode)
            builder.Builder.Append(" "c, repeatCount:=indent)
            builder.Builder.Append(node.Kind.ToString())
            If node.IsMissing Then
                builder.Builder.Append($" (missing)")
            End If
            builder.Builder.AppendLine()

            indent += 2
            MyBase.DefaultVisit(node)
            indent -= 2
        End Sub
    End Class
End Module

File: src/Compilers/Test/Utilities/CSharp/SyntaxTreeExtensions.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System;
using Microsoft.CodeAnalysis.CSharp.Symbols;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.PooledObjects;
using Microsoft.CodeAnalysis.Text;

namespace Microsoft.CodeAnalysis.CSharp.UnitTests
{
    public static class SyntaxTreeExtensions
    {
        public static SyntaxTree WithReplace(this SyntaxTree syntaxTree, int offset, int length, string newText)
        {
            var oldFullText = syntaxTree.GetText();
            var newFullText = oldFullText.WithChanges(new TextChange(new TextSpan(offset, length), newText));
            return syntaxTree.WithChangedText(newFullText);
        }

        public static SyntaxTree WithReplaceFirst(this SyntaxTree syntaxTree, string oldText, string newText)
        {
            var oldFullText = syntaxTree.GetText().ToString();
            int offset = oldFullText.IndexOf(oldText, StringComparison.Ordinal);
            int length = oldText.Length;
            return WithReplace(syntaxTree, offset, length, newText);
        }

        public static SyntaxTree WithReplace(this SyntaxTree syntaxTree, int startIndex, string oldText, string newText)
        {
            var oldFullText = syntaxTree.GetText().ToString();
            int offset = oldFullText.IndexOf(oldText, startIndex, StringComparison.Ordinal); // Use an offset to find the first element to replace at
            int length = oldText.Length;
            return WithReplace(syntaxTree, offset, length, newText);
        }

        public static SyntaxTree WithInsertAt(this SyntaxTree syntaxTree, int offset, string newText)
        {
            return WithReplace(syntaxTree, offset, 0, newText);
        }

        public static SyntaxTree WithInsertBefore(this SyntaxTree syntaxTree, string existingText, string newText)
        {
            var oldFullText = syntaxTree.GetText().ToString();
            int offset = oldFullText.IndexOf(existingText, StringComparison.Ordinal);
            return WithReplace(syntaxTree, offset, 0, newText);
        }

        public static SyntaxTree WithRemoveAt(this SyntaxTree syntaxTree, int offset, int length)
        {
            return WithReplace(syntaxTree, offset, length, string.Empty);
        }

        public static SyntaxTree WithRemoveFirst(this SyntaxTree syntaxTree, string oldText)
        {
            return WithReplaceFirst(syntaxTree, oldText, string.Empty);
        }

        internal static string Dump(this SyntaxNode node)
        {
            var visitor = new CSharpSyntaxPrinter();
            visitor.Visit(node);
            return visitor.Dump();
        }

        internal static string Dump(this SyntaxTree tree)
        {
            return tree.GetRoot().Dump();
        }

        private class CSharpSyntaxPrinter : CSharpSyntaxWalker
        {
            PooledStringBuilder builder;
            int indent = 0;

            internal CSharpSyntaxPrinter()
            {
                builder = PooledStringBuilder.GetInstance();
            }

            internal string Dump()
            {
                return builder.ToStringAndFree();
            }

            public override void DefaultVisit(SyntaxNode node)
            {
                builder.Builder.Append(' ', repeatCount: indent);
                builder.Builder.Append(node.Kind().ToString());
                if (node.IsMissing)
                {
                    builder.Builder.Append(" (missing)");
                }
                else if (node is IdentifierNameSyntax name)
                {
                    builder.Builder.Append(" ");
                    builder.Builder.Append(name.ToString());
                }
                builder.Builder.AppendLine();

                indent += 2;
                base.DefaultVisit(node);
                indent -= 2;
            }
        }
    }
}
